site/source/tutorials/1015_form_validation/index.md

```diff
@@ -0,0 +1,366 @@
+		novalidate: 'true',
+		onsubmit: this._onSubmit
+	}, [
+		v('fieldset', { classes: this.classes(css.nameField) }, [
+			v('legend', { classes: this.classes(css.nameLabel) }, [ 'Name' ]),
+			w(TextInput, {
+				key: 'firstNameInput',
+				label: {
+					content: 'First Name',
+					hidden: true
+				},
+				placeholder: 'Given name',
+				value: firstName,
+				required: true,
+				invalid: formErrors.firstName,
+				onInput: this.onFirstNameInput
+			}),
+			w(TextInput, {
+				key: 'lastNameInput',
+				label: {
+					content: 'Last Name',
+					hidden: true
+				},
+				placeholder: 'Surname name',
+				value: lastName,
+				required: true,
+				invalid: formErrors.lastName,
+				onInput: this.onLastNameInput
+			})
+		]),
+		w(TextInput, {
+			label: 'Email address',
+			type: 'email',
+			value: email,
+			required: true,
+			invalid: formErrors.email,
+			onInput: this.onEmailInput
+		}),
+		w(Button, {}, [ 'Save' ])
+	]);
+}
+```
+{% endsolution %}
+
+Now when you view the app in the browser, the border color of each form field changes as you type. Next we'll add error messages and update `onInput` validation to only occur after the first blur event.
+
+{% section %}
+
+## Extending TextInput
+
+{% task 'Create an error message' %}
+
+Simply changing the border color of form fields to be red or green doesn't impart much information to the user -- we need to add some error message text along with invalid state. On a basic level, our error text must be associated with a form input, styleable, and accessible. A single form field with an error message might look something like this:
+
+```ts
+v('div', { classes: this.classes(css.inputWrapper) }, [
+	w(TextInput, {
+		...
+		describedBy: this._errorId,
+		onInput: this._onInput
+	}),
+	invalid === true ? v('span', {
+		id: this._errorId,
+		classes: this.classes(css.error),
+		'aria-live': 'polite'
+	}, [ 'Please enter valid text for this field' ]) : null
+])
+```
+
+The error message is associated with the text input through `aria-describedby`, and the `aria-live` attribute ensures it will be read if it is added to the DOM or changed. Wrapping both the input and the error message in a containing `<div>` allows us to position the error message relative to the input if desired.
+
+{% instruction 'Extend `TextInput` to create a `ValidatedTextInput` widget with an error message and `onValidate` method' %}
+
+Re-creating the same error message boilerplate for multiple text inputs seems overly repetitive, so we're going to extend `TextInput` instead. This will also allow us to have better control over when validation occurs, e.g. by adding it to blur events as well. For now, just create a `ValidatedTextInput` widget that accepts the same properties interface as `TextInput` but with an `errorMessage` string and `onValidate` method. It should return the same node structure modeled above.
+
+{% solution showsolution4 %}
+```ts
+import { WidgetBase } from '@dojo/widget-core/WidgetBase';
```

Comment:
Missing instructions for the users to add the extra classes `inputWrapper` and `error` or should add them into the initial demo css for the user.